sky/backends/provision_utils.py

```diff
@@ -0,0 +1,494 @@
+        ['node_config'],
+        count=num_nodes,
+        tags={},
+        resume_stopped_nodes=True)
+
+    with _add_logger_handlers(log_abs_path):
+        try:
+            logger.debug(_TITLE.format('Provisioning'))
+            logger.debug('Provision config:\n'
+                         f'{json.dumps(bootstrap_config.dict(), indent=2)}')
+            return _bulk_provision(cloud, region, zones, cluster_name,
+                                   bootstrap_config)
+        except Exception:  # pylint: disable=broad-except
+            zone_str = 'all zones'
+            if zones:
+                zone_str = ','.join(zone.name for zone in zones)
+            logger.error(f'Failed to provision {cluster_name.display_name!r} '
+                         f'on {cloud} ({zone_str}).')
+            logger.debug(f'Starting instances for {cluster_name!r} '
+                         f'failed. Stacktrace:\n{traceback.format_exc()}')
+            # If cluster was previously UP or STOPPED, stop it; otherwise
+            # terminate.
+            # FIXME(zongheng): terminating a potentially live cluster is
+            # scary. Say: users have an existing cluster that got into INIT, do
+            # sky launch, somehow failed, then we may be terminating it here.
+            terminate = not is_prev_cluster_healthy
+            terminate_str = ('Terminating' if terminate else 'Stopping')
+            logger.error(f'{terminate_str} the failed cluster.')
+            teardown_cluster(repr(cloud),
+                             cluster_name,
+                             terminate=terminate,
+                             provider_config=original_config['provider'])
+            return None
+
+
+def teardown_cluster(cloud_name: str, cluster_name: ClusterName,
+                     terminate: bool, provider_config: Dict) -> None:
+    """Deleting or stopping a cluster."""
+    if terminate:
+        provision.terminate_instances(cloud_name, cluster_name.name_on_cloud,
+                                      provider_config)
+        metadata_utils.remove_cluster_metadata(cluster_name.name_on_cloud)
+    else:
+        provision.stop_instances(cloud_name, cluster_name.name_on_cloud,
+                                 provider_config)
+
+
+def _wait_ssh_connection_direct(
+        ip: str,
+        ssh_user: str,
+        ssh_private_key: str,
+        ssh_control_name: Optional[str] = None,
+        ssh_proxy_command: Optional[str] = None) -> bool:
+    del ssh_control_name
+    assert ssh_proxy_command is None, 'SSH proxy command is not supported.'
+    try:
+        with socket.create_connection((ip, 22), timeout=1) as s:
+            if s.recv(100).startswith(b'SSH'):
+                return True
+    except socket.timeout:  # this is the most expected exception
+        pass
+    except Exception:  # pylint: disable=broad-except
+        pass
+    logger.debug(f'Failed SSH to {ip}. Try: ssh -i {ssh_private_key} '
```

Comment:
Was poking around in #2507 and this seems terrifying... Somehow the debug level logs is still shown in the logs inside
```bash
I 09-08 10:44:14 cloud_vm_ray_backend.py:1429] To view detailed progress: tail -n100 -f /home/memory/sky_logs/sky-2023-09-08-10-44-07-687341/provision.log
```

```log
D 09-08 10:44:25 provision_utils.py:368] Waiting for SSH to be available for 'sky-3ce6-memory' ...
D 09-08 10:44:26 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:26 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:28 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:28 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:30 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:30 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:32 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:32 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:34 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:34 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:36 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:36 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:38 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:38 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:40 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:40 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:42 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:42 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:44 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:44 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:46 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:46 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:48 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:48 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:50 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:50 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:52 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:52 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:54 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:54 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:56 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:56 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:57 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:57 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:58 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:58 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:44:59 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:44:59 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:45:00 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:45:00 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:45:01 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:45:01 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:45:02 provision_utils.py:247] Failed SSH to 3.83.124.44. Try: ssh -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o ConnectTimeout=20s -o ubuntu@3.83.124.44 echo
D 09-08 10:45:02 provision_utils.py:321] Retrying in 1 second...
D 09-08 10:45:03 provision_utils.py:370] SSH Conection ready for 'sky-3ce6-memory'
```

Maybe we could use something like ray's `SSH Connection is not ready` instead of `Failed SSH to`?